phoenix-core/src/main/java/org/apache/phoenix/mapreduce/index/IndexUpgradeTool.java

```diff
@@ -318,53 +323,155 @@ public int executeTool() {
-                    if (!(test || dryRun || indexes.isEmpty())) {
-                        try {
-                            Thread.sleep((GLOBAL_INDEX_CHECKER_ENABLED_MAP_EXPIRATION_MIN + 1)
-                                    * 60 * 1000);
-                        } catch (InterruptedException e) {
-                            LOGGER.warning("Sleep before starting index rebuild interrupted. "
-                                    + e.getMessage());
-                        }
-                    }
-                }
                 enableTable(admin, dataTableFullName, indexes);
                 LOGGER.info("Completed " + operation + " of " + dataTableFullName);
-
-            } catch (IOException | SQLException e) {
-                LOGGER.severe("Something went wrong while executing " + operation
-                        + " steps " + e);
-                return -1;
             }
+        } catch (IOException | SQLException e) {
+            LOGGER.severe("Something went wrong while executing "
+                    + operation + " steps for "+ dataTableFullNameGlobal + " " + e);
+            handleFailure(queryServices, dataTableFullNameGlobal, mutableTables);
         }
         // Opportunistically kick-off index rebuilds after upgrade operation
-        if (upgrade) {
-            for (String dataTableFullName : tablesAndIndexes.keySet()) {
-                rebuildIndexes(conn, conf, dataTableFullName);
-                LOGGER.info("Started index rebuild post " + operation + " of "
-                        + dataTableFullName);
+        rebuildIndexes(conn, conf, mutableTables);
+    }
+
+    private void enableImmutableTables(ConnectionQueryServices queryServices,
+            ArrayList<String> immutableList,
+            long startWaitTime) {
+        int retryCount = 0;
+        long endWaitTime = System.currentTimeMillis();
+        long waitMore = getWaitMoreTime(endWaitTime, startWaitTime);
+        while  (retryCount < 5 && !waited) {
+            // If the table is immutable, we need to wait for clients to purge
+            // their caches of table metadata
+
+            LOGGER.info("waiting for more " + waitMore + " ms for client cache "
+                    + "to expire for immutable tables");
+            try {
+                startWaitTime = System.currentTimeMillis();
+                Thread.sleep(waitMore);
+                waited = true;
+            } catch (InterruptedException e) {
+                endWaitTime = System.currentTimeMillis();
+                waitMore = getWaitMoreTime(endWaitTime, startWaitTime);
+                retryCount++;
+                LOGGER.warning("Sleep before starting index rebuild is interrupted. "
+                        + "Attempting to sleep again! " + e.getMessage());
             }
         }
-        return 0;
+        if (retryCount == 5) {
+            LOGGER.info("Sleep was interrupted multiple times for immutable tables, "
```

Comment:
why do we eat this exception? Do we need the max retries since we already have a max timeout to prevent infinite loops?

Reply:
We want to retry in case of interruption during sleep. And only try a finite number of times. I think for this corner case scenario, we are okay if we see any problem with cache incoherency. What say?

Reply:
I agree with not having the needless retry-count logic. Whether the sleep is interrupted or not, as long as the clock moved beyond initial time + GLOBAL_INDEX_CHECKER_ENABLED_MAP_EXPIRATION_MIN + 1 minute, it shouldn't matter how many retries of sleep were needed. Safer than not sleeping enough and let caches be stale.

Reply:
I agree with @gjacoby126, no need to have the max retries here. I suggest removing retryCount and having 
while (waitMore  > 0) instead.

Reply:
Thank you, everyone! Initially, the logic wasn't there and I was concerned about stale cache then. Looks like we are okay there.